### PR TITLE
Advance Debugging

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- [x] https://github.com/theia-ide/theia/issues/5769 - watch expressions
  - [x] a CQ for add icon: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20445 (resolved)
- [x] https://github.com/theia-ide/theia/issues/5768 - exception breakpoints
- [x] https://github.com/theia-ide/theia/issues/5767 - function breakpoints
  - [x] a CQ for function breakpoint icon: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20470
- [x] https://github.com/theia-ide/theia/issues/3807 - inlined breakpoints
- [x] https://github.com/theia-ide/theia/issues/5788 - fix identify of breakpoints in the plugin process
- [x] add copyrights to all copied icons

Pause on an exception with watch expressions:
Screen Shot 2019-07-24 at 09 45 04

#### How to test

- watch expressions: add watch expressions and start debugging, they should be updated as you go
- exception breakpoints:
  - they are provided by a debug adapter on the first run, so start debugging something
  - after that you can enable them
  - after the debug session is finished, such breakpoints should be preserved anyway and reevaluated on next run
  - when exception breakpoint hits you should be able to see the reason in an editor
- function breakpoints:
  - install nodejs vscode extensions, https://marketplace.visualstudio.com/items?itemName=ms-vscode.node-debug2 and https://marketplace.visualstudio.com/items?itemName=ms-vscode.node-debug
  - uninstall native Theia nodejs extension from example app and rebuild it
  - follow https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_function-breakpoints
    - it works only with legacy protocol for Node.js < 8 and only for global functions which were already defined at runtime
function_breakpoint

- inlined breakpoints
  - look out for regressions with usual breakpoints, one should be able to mix them, disable, enable, edit on the same line via the context menu in the ruler and the breakpoints debug view

- breakpoints identity issues in the plugin host process: TBD

### CODE REVIEW

File: packages/core/src/browser/icons/add-inverse.svg [hunk 1]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><defs><style>.icon-canvas-transparent,.icon-vs-out{fill:#252526;}.icon-canvas-transparent{opacity:0;}.icon-vs-bg{fill:#c5c5c5;}</style></defs><title>add</title><g id="canvas"><path class="icon-canvas-transparent" d="M16,16H0V0H16Z"/></g><g id="outline" style="display: none;"><path class="icon-vs-out" d="M14,6v4H10v4H6V10H2V6H6V2h4V6Z"/></g><g id="iconBg"><path class="icon-vs-bg" d="M13,7V9H9v4H7V9H3V7H7V3H9V7Z"/></g></svg>
```
Comment (akosyakov) [inline packages/core/src/browser/icons/add-inverse.svg]: add copyrights to all copied icons

**Comment by tolusha** [comment]
@akosyakov Is it ready for review?

**Comment by akosyakov** [comment]
@tolusha I did not have time to add test sample VS Code extension to test identity of breakpoints in the plugin host process, but the rest should be good.

**Comment by tolusha** [comment]
I've testes with java [1] & java debug extensions [2]. Works fine:
- [x] watch expressions
- [x] exception breakpoints
Currently functional breakpoints are unsupported in java. Inline breakpoints doesn't work :(

[1] https://marketplace.visualstudio.com/_apis/public/gallery/publishers/redhat/vsextensions/java/0.47.0/vspackage
[2] https://marketplace.visualstudio.com/_apis/public/gallery/publishers/vscjava/vsextensions/vscode-java-debug/0.21.0/vspackage

**Comment by akosyakov** [comment]
@tolusha inline breakpoints don't hit with java debug VS Code extension? Do they hit with VS Code?

**Comment by tolusha** [comment]
@akosyakov 
you are right, Inline breakpoints are n't supported as well.
https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-debug
`Breakpoints/Conditional Breakpoints/Logpoints`

**Comment by tolusha** [comment]
I will do tests with some other plugins...

**Comment by tolusha** [comment]
@akosyakov 
I've tested Go [1] and Python [2] extensions and didn't find any difference in behavior between VS Code and Theia. If it works in VS Code then it work in Theia.

[1] https://marketplace.visualstudio.com/items?itemName=ms-vscode.Go 
[2] https://marketplace.visualstudio.com/items?itemName=ms-python.python

**Comment by thegecko** [comment]
Hi @akosyakov , do the exception breakpoints rely on the debug adapter supporting `supportsExceptionInfoRequest` (using the `exceptionInfoRequest()` function) to show the inline exception?

File: packages/debug/src/browser/editor/debug-editor-model.ts [hunk 2]
```diff
@@ -150,6 +150,26 @@ export class DebugEditorModel implements Disposable {
         return decorations;
     }
 
+    protected async toggleExceptionWidget(): Promise<void> {
+        const { currentFrame } = this.sessions;
+        if (!currentFrame ||
+            !currentFrame.source || currentFrame.source.uri.toString() !== this.uri.toString() ||
```
Comment (thegecko) [inline packages/debug/src/browser/editor/debug-editor-model.ts]: This shouldn't guard against the currentFrame not having source as it's common to have the topmost stack frame point at the exception and not have source:

response.body.stackFrames:

[
    {
        column: 0
        id: 1001
        line: 0
        name: "__cxa_throw"
        source: undefined
    },{
        column: 0
        id: 1002
        line: 24
        name: "main"
        source: Source {name: "main.cpp", path: ... }
    }
]

In this case it returns early and the exception isn't shown:

Screen Shot 2019-09-23 at 11 47 43

Comment (thegecko) [reply packages/debug/src/browser/editor/debug-editor-model.ts]: Unless the adapter is expected to strip that frame off, which sounds suspicious.

File: packages/debug/src/browser/editor/debug-editor-model.ts [hunk 3]
```diff
@@ -150,6 +150,26 @@ export class DebugEditorModel implements Disposable {
...
+            !currentFrame.raw.line || !currentFrame.raw.column) {
```
Comment (thegecko) [inline packages/debug/src/browser/editor/debug-editor-model.ts]: This code will fail if line or column are `0` (falsey values)

```suggestion
            currentFrame.raw.line === undefined || currentFrame.raw.column === undefined) {
```

**Comment by akosyakov** [comment]
@thegecko @tolusha thank you for review! I would try to find time to address feedback but probably only after EclipseCon.

If you are interested in some of these changes, feel free to extract it to separate PRs and land, i.e. without inline breakpoints but fixing other code issues.

**Comment by thegecko** [comment]
@akosyakov Are you planning to revisit this PR?

**Comment by akosyakov** [comment]
@thegecko I'm afraid not in this year :( But anyone can do it, feel free to rebase it and continue in the same branch or extract some parts in new PRs.

**Comment by tolusha** [comment]
@thegecko 
Is there any chance you have started working on this PR?
I would like to rebase and address remaining issues.

**Comment by akosyakov** [comment]
@tolusha please do, as usual original commits and authorship should be preserved, but feel free to rebase them and push additional commits to fix remaining issues

**Comment by thegecko** [comment]
> Is there any chance you have started working on this PR?

Not yet, its about number 10 on my todo list :)

File: packages/core/src/browser/icons/add-inverse.svg (hunk 1 above)
Comment (tolusha) [reply packages/core/src/browser/icons/add-inverse.svg]: @akosyakov 
Where did you copy them from? VS code?

**Comment by tolusha** [comment]
@thegecko 

> Nit: the + button still appears for adapters which don't support supportsFunctionBreakpoints
> Log Points always appear in the UI even if unsupported in the adapter (supportsLogPoints)
> Hit Count always appear in the UI even if unsupported in the adapter (supportsHitConditionalBreakpoints)

VS Code behaves in the same way. It probably comes from the fact that when you edit breakpoint VS  Code doesn't know which debugger is supposed to be used.

**Comment by thegecko** [comment]
> VS Code behaves in the same way

That's strange because the [capabilities](https://microsoft.github.io/debug-adapter-protocol/specification#Types_Capabilities) outline what the adapter can support.

Anyway, we should mimic VSCode in this respect I guess?

**Comment by tolusha** [comment]
@thegecko
I've rebased the branch.
Inline breakpoints should be available throw menu `debug->new breakpoints->inline breakpoint`

**Comment by tolusha** [comment]
VS Code allows to add all breakpoints by notify user if they aren't supported
image

File: packages/debug/src/browser/editor/debug-editor-service.ts [hunk 4]
```diff
@@ -106,8 +106,8 @@ export class DebugEditorService {
         return this.model && this.model.getBreakpoint(position);
     }
 
-    get inlineBreakpoint(): DebugSourceBreakpoint | undefined {
-        return this.model && this.model.inlineBreakpoint;
+    inlineBreakpoint(position?: monaco.Position): DebugSourceBreakpoint | undefined {
```
Comment (akosyakov) [inline packages/debug/src/browser/editor/debug-editor-service.ts]: `getInlineBreakpoint`

**Comment by akosyakov** [comment]
> That's strange because the capabilities outline what the adapter can support.

@thegecko capabilities are available only after a user run at least something at least once, otherwise there were no chance to fetch them, but still a user should be able to add exception/functional breakpoints initially

**Comment by tolusha** [comment]
@akosyakov 
Regarding adding license to all copied icons.
Are we talking about only `add.svg` && `add-inverse.svg` or about all icons inside `icons` folder?
Did you copied them from VS Code?

**Comment by tolusha** [comment]
It is ready to test.

File: packages/debug/src/browser/style/breakpoint-function-disabled.svg [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (akosyakov) [inline packages/debug/src/browser/style/breakpoint-function-disabled.svg]: @tolusha I've copied breakpoing svg files also from VS Code. CQs were approved already, but please add copyrights to them 🙏

**Comment by akosyakov** [comment]
I've removed management of inline breakpoins from the context menu. We can do it later. I don't have time to test unfortunately.

File: packages/debug/src/browser/style/breakpoint-function-disabled.svg (hunk 5 above)
Comment (tolusha) [reply packages/debug/src/browser/style/breakpoint-function-disabled.svg]: Done

**Comment by akosyakov** [comment]
We need to apply following colors as soon as the theming PR is in:

> The Debug Exception widget is a peek view that shows in the editor when debug stops at an exception.
debugExceptionWidget.background: Exception widget background color.
debugExceptionWidget.border: Exception widget border color.

from https://code.visualstudio.com/api/references/theme-color#editor-widget-colors

I don't think we want this PR in for next release. We want to finish theming, merging it will require additional work. I would prefer to add couple styles later, beginning Jan and then test/merge it.

**Comment by tolusha** [comment]
@akosyakov 
I probably won't be able handle this PR starting next week

**Comment by akosyakov** [comment]
@tolusha the problem is that we are going to prohibit to hard code colors in css, only css color variables can be used which are first registered via `ColorContribution`, i.e. all colors should be themable. This PR introduces such hardcoding.

You can change base of this PR to branch of https://github.com/eclipse-theia/theia/pull/6722, introduce new colors and update css to use them. So this PR will be prepared for testing after theming PR is merged. Colors should be registered with ids as in VS Code, i.e. with `debugExceptionWidget.background` and `debugExceptionWidget.border`. For default values and descriptions please look up VS Code code base. After registration colors can be referenced in css with `var(--theia-debugExceptionWidget-background)` and `var(--theia-debugExceptionWidget-border)`.

**Comment by akosyakov** [comment]
I'm going to reset base branch for it as https://github.com/eclipse-theia/theia/pull/6722 and add missing colors.

**Comment by akosyakov** [comment]
- [ ] In the meantime new colors were added for debug icons: https://code.visualstudio.com/api/references/theme-color#debug-icons

**Comment by akosyakov** [comment]
I've fixed theming, so it should be fine to merge after the theming PR is in.

@thegecko I think all comments except using the context menu to add inline breakpoints were fixed. Would it be alright to tackle management of inline breakopints via the context menu separately? Right now one can add it by putting a cursor at some position and using the command palette, the debug top-level menu or the short-cut `shift + f9`.

@tolusha If you are fine with changes please don't forget to give an approve.

**Comment by thegecko** [comment]
@akosyakov, perhaps open an issue to add the context menu item? Happy to merge as is.

**Comment by akosyakov** [comment]
@thegecko I've filed an issue for inline breakpoints management with the context menu: https://github.com/eclipse-theia/theia/issues/6795 I'm going to rebase the PR and if the build is green then merge it.

**Pull request merged**